lib/encoder/src/Pololu37D.cpp

```diff
@@ -10,18 +10,46 @@ Pololu37D::Pololu37D(const Config &config)
 
 bool Pololu37D::getAngleDeg(float &angle) {
   std::scoped_lock<Mutex> lock(m_mutex);
-  angle = (m_QEI.getPulses() * m_degreesPerCount) - m_zeroOffsetDeg;
+  read();
+  angle = m_current_angle_deg;
   return true;
 }
 
 bool Pololu37D::getAngularVelocityDegPerSec(float &speed) {
   std::scoped_lock<Mutex> lock(m_mutex);
-  speed = m_QEI.getPulseVelocity_PulsesPerSec() * m_degreesPerCount;
+  read();
+  speed = m_speed_deg_per_second;
   return true;
 }
 
 bool Pololu37D::reset() {
   std::scoped_lock<Mutex> lock(m_mutex);
+  m_zeroOffsetDeg        = 0.0;
+  m_current_angle_deg    = 0.0;
+  m_previous_angle_deg   = 0.0;
+  m_delta_time_ns        = 0.0;
+  m_speed_deg_per_second = 0.0;
   m_QEI.reset();
   return true;
 }
+
+float Pololu37D::return_delta_time() {
+  return m_delta_time_ns;
+}
+
+bool Pololu37D::read() {
+  // time since last measurement
+  m_timer.stop();
+  m_delta_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(m_timer.elapsed_time()).count();
+  m_timer.reset();
+  m_timer.start();
+
+  m_previous_angle_deg = m_current_angle_deg;
+
+  m_current_angle_deg = (m_QEI.getPulses() * m_degreesPerCount) - m_zeroOffsetDeg;
+
+  // to estimate speed: delta_angle / delta_time (degrees / sec)
+  m_speed_deg_per_second = (m_current_angle_deg - m_previous_angle_deg) / m_delta_time_ns * 1000000000;
```

Comment:
What does the very first ang velocity reading give? You might want to add a divide by zero check on dt, or some special case for ang velocity if it's the first reading

Reply:
It returns a speed of 0. I have added those checks. Currently, I'm going to have it print an error message let me know if I should make it do something else.

Reply:
![image](https://user-images.githubusercontent.com/50745474/110031330-c72eb400-7d04-11eb-8e4e-0066cdda7a14.png)

Now it looks like the first reading always returns an error. Is this going to be an issue?

Reply:
You should have some special handling for the first velocity reading to make sure that it doesn't give a read error. I think returning velocity = 0 is fine for the very first reading

Reply:
I've implemented that let me know what you think